networkit/cpp/graph/Graph.cpp

```diff
@@ -919,16 +919,48 @@ std::pair<node, node> Graph::randomEdge(bool uniformDistribution) const {
 	}
 
 	if (uniformDistribution) {
-		return randomEdges(1)[0];
+		/*
+		 * The simple idea here is to interpret all neighborhoods next to each other, resulting
+		 * in a virtual vector of size m. Then we draw a random index and return the edge.
+		 * For undirected edges, the vector has size 2m; but the idea remains. There is one minor
+		 * complication for undirected edges with self-loops: each edge {u,v} with u != v is stored
+		 * twice (once in the neighborhood of u, once in v) but a loop (u, u) is only stored once.
+		 * To equalize the probabilities we reject edges {u,v} with u > v and try again. This leads
+		 * to less than two expected trails in and is only done for undirected graphs with self-loops.
+		 */
+
+		while(true) {
+			node u = 0;
+
+			const auto upper = directed ? numberOfEdges() : 2 * numberOfEdges() - numberOfSelfLoops();
+			auto idx = Aux::Random::index(upper);
+
+			while (idx >= degreeOut(u)) {
```

Comment:
Would that be possible to initialize another node to `upperNodeIdBound() - 1`, another index to `numberOfEdges() - idx`, and iterate also from above? This could reduce the number of iterations of this loop.

Reply:
If we assume that degrees are randomly distributed, your suggestion works. However, I often deal with graph in which this is not the case and then your suggestion would be pessimization. Hence, as the outcome is unclear and we just got a speed-up of one order of magnitude, I would lean to not implement it for clarity's sake.

If you feel strongly about it, I obviously will do it.

Reply:
What about the following compromise: if idx < m / 2 start from begin, if idx > m/2 start from end?

Reply:
Then I think that it is better to leave it unchanged.

Reply:
If we do not have any prior information about the degree distribution, I guess that this is the best that we can do. So I agree with this compromise.